File: code/matlab/preprocessing/HMInputsPreparation.m
%% @ Lingfeng Zhou 
%% generate input files for SWAT-HM
% icc: control code for for HM inputs files
% 1. metal.dat
% 2. .hml files: hru level inputs and parameters for metals
% 3. .swq files: river channel level and parameters for metals
% 4. ***p.dat files: point source files

%% Define inputs
clc
clear
promdir = pwd;
xlsdir = [promdir '\HeavyMetalModuleDataBase_XiangRiver_XRB210528V2.xlsx']; % database excel file directory
% projdir = 'D:\XiangRiverProject\XiangRiver\XRB210528V2.Sufi2.SwatCup'; % folder of SWAT-HM project
projdir = 'D:\XiangRiverProject\XiangRiver\XRB210528V2_1.Sufi2.SwatCup'; % folder of SWAT-HM project
[iprint,nyskip,SubNo,HruNo] = readfilecio(projdir);
icc=[1,0,1,0]; % icc: control code for for HM inputs files, 1=yes, 0=no
PointSourceNo=[1:SubNo]; % if icc(4)=1, then PointSourceNo is the No of point source files

tic
for ii=1:numel(icc)
    %% metal.dat file
    if (ii==1 && icc(ii)==1)
        % change the range of excel sheet in your case
        [~, ~, HmlPar] = xlsread(xlsdir,'Hml_Prameters','A2:O2');
        filename = [projdir '\metal.dat'];
        fid=fopen(filename,'w');
        formatSpec = '%3i%10s%3i%10.3f%10.3f%10.3f%10.6f%10.6f%10.3f%10.6f%10.6f%10.6f%10.3f%10.6f%10.6f\n\r';
        fprintf(fid, formatSpec, HmlPar{1:15});
        fclose(fid);
        disp('1 metal.dat file is modified')
        clearvars fid filename formatSpec
    end
    %% .hml files
    if (ii==2 && icc(ii)==1)
        % change the range of excel sheet in your case
        [num, ~, ~] = xlsread(xlsdir,'Hml_Hru');
        HmlHru = num(:,2:11);
        
        filenames = dir([projdir '\0*.hru']);
        for jj=1:numel(filenames)
            hrufilename=filenames(jj).name;
            hmlfilename=[hrufilename(1:9), '.hml'];
            hrufileid=fopen([projdir '\' hrufilename],'r');
            tline = fgetl(hrufileid);
            fclose(hrufileid);
            
            hmlfileid=fopen([projdir '\' hmlfilename],'w');
            fprintf(hmlfileid,[' .hml' tline(6:end),'\n']);
            fprintf(hmlfileid,'Soil Heavy Metal Data\n');
            fprintf(hmlfileid,'%10d    |HML_P1: Metal number [#]\n',1);
            fprintf(hmlfileid,'%10.6f    |HML_P2: Heavy metal Nonpoint source Area Fraction [-]\n',HmlHru(jj,1));
            fprintf(hmlfileid,'%10.3f    |HML_P3: Heavy metal in Rock [kg/ha]\n',HmlHru(jj,2));
            fprintf(hmlfileid,'%10.3f    |HML_P4: Labile metal in 1st layer soil [mg/kg]\n',HmlHru(jj,3));
            fprintf(hmlfileid,'%10.3f    |HML_P5: Non-labile metal in 1st layer soil [mg/kg]\n',HmlHru(jj,4));
            fprintf(hmlfileid,'%10.3f    |HML_P6: Enrichment ratio of heavy metal [-]\n',HmlHru(jj,5));
            %fprintf(hmlfileid,'%10.3f    |HML_P7: Soil pH [-]\n',HmlHru(jj,7));
            fprintf(hmlfileid,'%10.3f    |HML_P8: Total metal input from agricultural use [g/ha/yr]\n',HmlHru(jj,6));
            fprintf(hmlfileid,'%10.3f    |HML_P9: Fraction of labile metal in fertilizers or animal manure [-]\n',HmlHru(jj,7));
            fprintf(hmlfileid,'%10.3f    |HML_P10: Total metal input from atmospheric deposition [g/ha/yr]\n',HmlHru(jj,8));
            fprintf(hmlfileid,'%10.3f    |HML_P11: Fraction of labile metal in atmospheric deposition [-]\n',HmlHru(jj,9));
            fprintf(hmlfileid,'%10.3f    |HML_P12: Dissolved metal in groundwater [ug/L]\n',HmlHru(jj,10));
            fclose(hmlfileid);
        end
        disp([num2str(jj) ' .hml files are modified'])
        clearvars num filenames hmlfilename hrufilename hmlfileid hrufileid tline
    end
    %% .swq files
    if (ii==3 && icc(ii)==1)
        [num, ~, ~] = xlsread(xlsdir,'Hml_Swq');
        HmlSwq = num;
        
        filenames = dir([projdir '\0*.swq']);
        for jj=1:numel(filenames)
            filename=filenames(jj).name;
            swqfileid=fopen([projdir '\' filename],'r+');
            for kk=1:30
                headline=fgetl(swqfileid);
            end
            fseek(swqfileid,0,'cof');
            fprintf(swqfileid,'Heavy Metal Parameters: \n');
            fprintf(swqfileid,'     %10.4f     | SWQ_HML_STL: Settling velocity for Heavy Metal [m/d]\n',HmlSwq(jj,2));
            fprintf(swqfileid,'     %10.4f     | SWQ_HML_RSP: Resuspension velocity for Heavy Metal [m/d]\n',HmlSwq(jj,3));
            fprintf(swqfileid,'     %10.4f     | SWQ_HML_MIX: Mixing velocity for Heavy Metal [m/d]\n',HmlSwq(jj,4));
            fprintf(swqfileid,'     %10.5f     | SWQ_HML_BRY: Burial velocity for Heavy Metal [m/d]\n',HmlSwq(jj,5));
            fprintf(swqfileid,'     %10.4f     | SWQ_HML_LabileCONC: Initial HM concentration in reach bed sediment [kg/m3]\n',HmlSwq(jj,6));
            fprintf(swqfileid,'     %10.4f     | SWQ_HML_NonLabileCONC: Initial HM concentration in reach bed sediment [kg/m3]\n',HmlSwq(jj,7));
            fprintf(swqfileid,'     %10.4f     | SWQ_HML_ACT: Depth of active sediment layer for heavy metal [m]\n',HmlSwq(jj,8));
            fclose(swqfileid);
        end
        disp([num2str(jj) ' .swq files are modified'])
        clearvars kk num filenames filename headline swqfileid
    end
    %% point source file
    if (ii==4 && icc(ii)==1)
        [num, ~, ~] = xlsread(xlsdir,'PointSource');
         HmlPoint = num;
        for jj=1:numel(PointSourceNo)

            psfilename = [projdir '\' num2str(PointSourceNo(jj)) 'p.dat'];
            Notes={'2016/4/21 0:00:00 .dat file Daily Record Subbasin  10 ArcSWAT 2012.10_2.16 interface';
                '    ';
                '    ';
                '    ';
                '    ';
                ' DAY YEAR          FLOCNST          SEDCNST         ORGNCNST         ORGPCNST          NO3CNST          NH3CNST          NO2CNST         MINPCNST         CBODCNST        DISOXCNST         CHLACNST       SOLPSTCNST       SRBPSTCNST        BACTPCNST       BACTLPCNST        CMTL1CNST        CMTL2CNST        CMTL3CNST';
                };
            
            psfileid=fopen(psfilename,'w');
            for kk=1:numel(Notes)
                fprintf(psfileid,[Notes{kk} '\n']);
            end
            formatSpec='%4s%5s %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E %16.3E\n';
            for kk=1:numel(HmlPoint(:,1))
                fprintf(psfileid,formatSpec,num2str(HmlPoint(kk,1)),num2str(HmlPoint(kk,2)),...
                    HmlPoint(kk,jj+2)*50000,zeros(1,14),HmlPoint(kk,jj+2),zeros(1,2));
            end
            fclose(psfileid);
        end
        disp([num2str(numel(PointSourceNo)) ' point source files are modified'])
        clearvars kk num psfilename psfileid formatSpec Notes
    end
end
toc

File: code/python/preprocessing/HMInputsPreparation.py
# -*- coding: utf-8 -*-

import pandas as pd
import os

pd.set_option('display.notebook_repr_html',False)


def get_base_num():
    sub_num = 0
    hru_num = 0
    file_list = os.listdir(PRO_DIR)
    for file_name in file_list:
        if file_name.startswith("0"):
            if file_name.endswith(".sub"):
                sub_num += 1
            if file_name.endswith(".hru"):
                hru_num += 1
    return sub_num, hru_num


PRO_DIR = "../../../data/TxtInOut2"
INPUT_FILE_DIR = '../../../data/HeavyMetalModuleDataBase2.xls'
SUB_NUM, HRU_NUM = get_base_num()


def write_file(filedir,filepath,lines,op_type):
    with open(f"{filedir}/{filepath}",op_type) as f:
        for l in lines:
            f.write(l)


def hml_parm(table_info):
    mat_dat = dict(table_info["Hml_Prameters"])
    mat_dat_str = "".join([mat_dat["ID"][0].astype(str).rjust(3),
           mat_dat["Name"][0].rjust(10),
           mat_dat["Eqn"][0].astype(str).rjust(3),
           round(mat_dat["Kd1"][0],3).astype(str).rjust(10),
           round(mat_dat["Kd2"][0],3).astype(str).rjust(10),
           round(mat_dat["Kd3"][0],3).astype(str).rjust(10),
           round(mat_dat["k1"][0],6).astype(str).rjust(10),
           round(mat_dat["k-1"][0],6).astype(str).rjust(10),
           round(mat_dat["ksol"][0],3).astype(str).rjust(10),
           round(mat_dat["kweth"][0],6).astype(str).rjust(10),
           round(mat_dat["ku"][0],6).astype(str).rjust(10),
           round(mat_dat["gamma"][0],6).astype(str).rjust(10),
           round(mat_dat["ka"][0],3).astype(str).rjust(10),
           round(mat_dat["kd"][0],3).astype(str).rjust(10),
           round(mat_dat["kwash"][0],3).astype(str).rjust(10),]) + "\r\n"
    print('1 metal.dat file is modified')
    write_file(PRO_DIR,"mat.dat",[mat_dat_str],"w")


def hml_hru(table_info):
    hml_data = dict(table_info["Hml_Hru"])
    hru_list = []
    for filename in os.listdir(PRO_DIR):
        if filename.endswith(".hru") and filename.startswith("0"):
            hru_list.append(str(filename[:9]))

    for i in range(len(hru_list)):
        hru_id = hru_list[i]
        hml_lines = []
        hruf = open(f"{PRO_DIR}/{hru_id}.hru","r")
        hrus = hruf.readlines()[0][6:]

        hml_lines.append(f" .hml {hrus}")
        hml_lines.append("Soil Heavy Metal Data\n")
        hml_lines.append(f"{'1'.rjust(10)}    |HML_P1: Metal number [#]\n")
        hml_lines.append(f"{round(hml_data['hmfraction'][i],6).astype(str).rjust(10)}    |HML_P2: Heavy metal Nonpoint source Area Fraction [-]\n")
        hml_lines.append(f"{round(hml_data['hmrock'][i],3).astype(str).rjust(10)}    |HML_P3: Heavy metal in Rock [kg/ha]\n")
        hml_lines.append(f"{round(hml_data['Labile hm concentration'][i],3).astype(str).rjust(10)}    |HML_P4: Labile metal in 1st layer soil [mg/kg]\n")
        hml_lines.append(f"{round(hml_data['Non-labile hm concentration'][i],3).astype(str).rjust(10)}    |HML_P5: Non-labile metal in 1st layer soil [mg/kg]\n")
        hml_lines.append(f"{round(hml_data['Enrichment ratio '][i],3).astype(str).rjust(10)}    |HML_P6: Enrichment ratio of heavy metal [-]\n")
        hml_lines.append(f"{round(hml_data['Total hm in Fertilizers'][i],3).astype(str).rjust(10)}    |HML_P8: Total metal input from agricultural use [g/ha/yr]\n")
        hml_lines.append(f"{round(hml_data['Fraction of labile hm in Fertilizers'][i],3).astype(str).rjust(10)}    |HML_P9: Fraction of labile metal in fertilizers or animal manure [-]\n")
        hml_lines.append(f"{round(hml_data['Total hm in atmospheric deposition'][i],3).astype(str).rjust(10)}    |HML_P10: Total metal input from atmospheric deposition [g/ha/yr]\n")
        hml_lines.append(f"{round(hml_data['Fraction of labile hm in atmospheric deposition'][i],3).astype(str).rjust(10)}    |HML_P11: Fraction of labile metal in atmospheric deposition [-]\n")
        hml_lines.append(f"{round(hml_data['Dissolved hm in groundwater'][i],3).astype(str).rjust(10)}    |HML_P12: Dissolved metal in groundwater [ug/L]\n")
        write_file(PRO_DIR, f"{hru_id}.hml", hml_lines, "w")
        print(f"{hru_id}.hml files are modified")

def hml_swq(table_info):
    swq_data = dict(table_info["Hml_Swq"])
    swq_list = []
    for filename in os.listdir(PRO_DIR):
        if filename.endswith(".swq") and filename.startswith("0"):
            swq_list.append(filename)

    for i in range(len(swq_list)):
        swq_filename = swq_list[i]
        swqf = open(f"{PRO_DIR}/{swq_filename}","r",encoding="ISO-8859-1")
        swqs = swqf.readlines()[:30]
        swqs.append("Heavy Metal Parameters: \n")
        swqs.append(f"     {round(swq_data['HML_STL'][i],4).astype(str).rjust(10)}     | SWQ_HML_STL: Settling velocity for Heavy Metal [m/d]\n")
        swqs.append(f"     {round(swq_data['HML_RSP'][i],4).astype(str).rjust(10)}     | SWQ_HML_RSP: Resuspension velocity for Heavy Metal [m/d]\n")
        swqs.append(f"     {round(swq_data['HML_MIX'][i],4).astype(str).rjust(10)}     | SWQ_HML_MIX: Mixing velocity for Heavy Metal [m/d]\n")
        swqs.append(f"     {round(swq_data['HML_BRY'][i],4).astype(str).rjust(10)}     | SWQ_HML_BRY: Burial velocity for Heavy Metal [m/d]\n")
        swqs.append(f"     {round(swq_data['LabileHML_CONC'][i],4).astype(str).rjust(10)}     | SWQ_HML_LabileCONC: Initial HM concentration in reach bed sediment [kg/m3]\n")
        swqs.append(f"     {round(swq_data['NonLabileHML_CONC'][i],4).astype(str).rjust(10)}     | SWQ_HML_NonLabileCONC: Initial HM concentration in reach bed sediment [kg/m3]\n")
        swqs.append(f"     {round(swq_data['HML_ACT'][i],4).astype(str).rjust(10)}     | SWQ_HML_ACT: Depth of active sediment layer for heavy metal [m]\n")
        write_file(PRO_DIR, swq_filename, swqs, "w")
        print(f"{swq_filename} files are modified")

def point_source(table_info):
    pointSource = table_info["PointSource"].values
    # print(pointSource)
    for i in range(1,SUB_NUM):
        psfilename = f"{i}p.dat"
        pslines = []
        pslines.append("2016/4/21 0:00:00 .dat file Daily Record Subbasin  10 ArcSWAT 2012.10_2.16 interface\n")
        pslines.append("    \n")
        pslines.append("    \n")
        pslines.append("    \n")
        pslines.append("    \n")
        pslines.append(" DAY YEAR          FLOCNST          SEDCNST         ORGNCNST         ORGPCNST          NO3CNST          NH3CNST          NO2CNST         MINPCNST         CBODCNST        DISOXCNST         CHLACNST       SOLPSTCNST       SRBPSTCNST        BACTPCNST       BACTLPCNST        CMTL1CNST        CMTL2CNST        CMTL3CNST\n")
        pslen = len(pointSource[0])
        for l in range(1, pslen):
            psinfo = [f"{str(pointSource[l][0].astype(int)).rjust(4)} {str(pointSource[l][1].astype(int)).rjust(4)}",
                      f"{(pointSource[l][i+2]*50000):.3E}".rjust(16)] + [f"{0:.3E}".rjust(16) for i in range(16)] + \
                     [f"{(pointSource[l][i+2]):.3E}".rjust(16)] + \
                     [f"{0:.3E}".rjust(16) for i in range(2)]
            pslines.append(" ".join(psinfo) + "\n")
        write_file(PRO_DIR, psfilename, pslines, "w")


# 读取xls（绝对路径）
table_info = pd.read_excel(io=INPUT_FILE_DIR,sheet_name=None)

handlers = ["hml_parm","hml_hru","hml_swq","point_source"]
for handler in handlers:
    if handler == "hml_parm":
        hml_parm(table_info)
    if handler == "hml_hru":
        hml_hru(table_info)
    if handler == "hml_swq":
        hml_swq(table_info)
    if handler == "point_source":
        point_source(table_info)
